crates/store/src/state.rs

```diff
@@ -471,20 +478,10 @@ impl State {
 
 /// Creates a [BlockNoteTree] from the `notes`.
 #[instrument(target = "miden-store", skip_all)]
-pub fn build_note_tree(notes: &[NoteCreated]) -> Result<BlockNoteTree, ApplyBlockError> {
-    // TODO: create SimpleSmt without this allocation
-    let mut entries: Vec<(usize, usize, (RpoDigest, NoteMetadata))> =
-        Vec::with_capacity(notes.len() * 2);
-
-    for note in notes.iter() {
-        let note_metadata =
-            NoteMetadata::new(note.sender.try_into()?, note.note_type, note.tag.into(), ZERO)?;
-        entries.push((
-            note.batch_index as usize,
-            note.note_index as usize,
-            (note.note_id, note_metadata),
-        ));
-    }
+pub fn build_note_tree<'a>(
+    notes: impl Iterator<Item = (BlockNoteIndex, &'a OutputNote)>,
+) -> Result<BlockNoteTree, ApplyBlockError> {
+    let entries = notes.map(|(note_index, note)| (note_index, note.id().into(), *note.metadata()));
 
     BlockNoteTree::with_entries(entries).map_err(ApplyBlockError::FailedToCreateNoteTree)
 }
```

Comment:
I wonder if this should be a method on the `Block` struct - e.g., something like:
```Rust
pub fn build_note_tree(&self) -> BlockNoteTree
```

Reply:
Got it, here is PR: https://github.com/0xPolygonMiden/miden-base/pull/630